### ISSUE TITLE
Code execution stops at invisible debug point 

### ISSUE THREAD

**Issue by sanjmeh** [rstudio/rstudio#15016]
### System details

  RStudio 2024.04.2+764 "Chocolate Cosmos" Release (e4392fc9ddc21961fd1d0efd47484b43f07a4177, 2024-06-05) for Ubuntu Jammy
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36, Quarto 1.4.555
 R Version, nickname:  Race for Your Life 

### Steps to reproduce the problem

- Create a debug point inside a function called by another function.
- Source this file.
- Call the original function from console.
- Code execution stops at the debug point inside the function called. (this is expected)
- Now uncheck the debug point (such that no red dot is visible)
- Source the target function file.
- Press up arrow and execute the last line of code in console.
- Code execution still stops at the same line (not expected).

### Describe the problem in detail
Code execution breaks when a break point is added inside a called function in another source file. Rstudio has a great feature of adding or removing a break point even while stepping through code. However at times (this is not clear when) the code breaks at the point even after debug point has been removed.

### Describe the behaviour you expected
As soon as a debug point is removed (by toggling debug points, through tools menu or just mouse click on red dot) the next execution of the function should not break at the point.

- [ x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [ x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.posit.co/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [ x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@sanjmeh Thank you for raising the issue! I'm unable to reproduce it, though. Can you provide a minimal reprex, with code sample with two functions? Thanks!

**Comment by sanjmeh** [comment]
> @sanjmeh Thank you for raising the issue! I'm unable to reproduce it, though. Can you provide a minimal reprex, with code sample with two functions? Thanks!

I will need to replicate this by trial and error. Will wait for the next occurrence.

**Comment by ronblum** [comment]
@sanjmeh Does this occur only with some code, or does it occur intermittently with the same code?

**Comment by sanjmeh** [comment]
> @sanjmeh Does this occur only with some code, or does it occur intermittently with the same code?

I have found it to come up intermittently with all code. Perhaps one common factor is when nested calls of functions happen.

**Comment by shirewoman2** [comment]
I've found this to happen intermittently as well. I don't know how to predict it, and I can't offer reprex precisely because it's so unpredictable. I have also found that sometimes RStudio will enter debug mode at a red dot in a function in a package I'm building and sometimes it does not, and I have no idea what makes it work and then not work. The same code will stop one session and then, a different day, it will not.

**Comment by jthomasmock** [comment]
We're having trouble reproducing this one -- it would be great if we could have one (mostly) reproducible example of code even if intermittently failing.

**Comment by Manjimadutta** [comment]
I encountered a related situation few days ago.

It happened the other way around, with the red dot on (at least it was visible on the right side whitespace), but the option 'Clear All Breakpoints' under the 'Debug' tool was unclickable, implying that it couldn't identify any debug/break points as such.

It causes chaos when you try to remove all red dots at once using this option, but it is jammed and won't let you; instead, we have to go line by line and remove them manually.

**Comment by olivroy** [comment]
I think I just encountered this.

In a version control repo (package),

1. set a break point in a function
2. Call that function
3. Modify the code slightly
4. Go to the git pane
5. Revert changes in to that line
6. load_all()
7. Call the function

Or maybe when the same project is open in 2 sessions, it can happen too. I think it eventually solves itself..

**Comment by olivroy** [comment]
In R 4.4, it doesn't work very smoothly, maybe some code related to https://github.com/rstudio/rstudio/pull/14452 should be adjusted?

**Comment by olivroy** [comment]
FWIW, the other issue I was encountering has been fixed in the latest dailies (#15072), but I am still encountering this one!

**Comment by kevinushey** [comment]
@olivroy do you have a reproducible example?

**Comment by olivroy** [comment]
@kevinushey Unfortunately, I don't really have one. It seems that it is a breaking point that has been set in a previous session, I probably switched branches somewhere with that file open. Is there a specific location in .Rproj.user I should look at where breakpoints are stored? I am pretty sure that if I deleted the .Rproj.user folder, the breakpoint would go away.. Is it safe to delete that folder?

**Comment by jthomasmock** [comment]
Yes you can delete the `.Rproj.user` folder -- let us know if that resolves the issue.

**Comment by olivroy** [comment]
It is strange. I tried deleting 

I have this in the .Rproj.user/2B6CC96E/pcs/debug-breakpoints.pper


```
{
    "debugBreakpointsState": {
        "breakpoints": [
            {
                "id": 15,
                "path": "~/rrr-forks/gt/R/resolver.R",
                "line_number": 300,
                "function_steps": "5,3,2",
                "function_name": "resolve_cols_i",
                "state": 1,
                "editor_state": 1,
                "editor_line_number": 300,
                "is_pending_debug_completion": false,
                "needs_updated_steps": false,
                "type": 0,
                "is_package_breakpoint": true,
                "package_name": "gt"
            }
        ]
    }
}
```

If I delete this file (or its contents or the entire .Rproj.user directory), it ends up reappearing, and I am still stuck with the same problem.

**Comment by olivroy** [comment]
@kevinushey I stumbled on this again. Would it be possible to have a fix to make sure that the Clear all breakpoints called from the command palette destroys completely  .Rproj.user/2B6CC96E/pcs/debug-breakpoints.pper. This way, we'd avoid having ghost breakpoints. 

The workaround to delete this file manually doesn't work well, because you have to quit the R session to be able to delete the file. If I delete it while the Rproject is open. RStudio will instantly write it back.
